keps/sig-api-machinery/0034-20190129-kflag.md

```diff
@@ -0,0 +1,188 @@
+  [implicitly registering flags](https://github.com/kubernetes/kubernetes/pull/57613)
+  against a component's command-line interface, which is required for components to maintain
+  explicit control over their command-line interface.
+
+Rather than require other components to copy and customize the Kubelet's relatively complex
+solutions to these problems, we should put the common functionality in a library, kflag, to reduce
+the burden on component owners.
+
+### Goals
+
+Provide core Kubernetes components with common building blocks that simplify the following:
+* Application of parsed flag values to arbitrary structs (e.g. config structs).
+* Selective inclusion of globally-registered flags in third-party code.
+
+### Non-Goals
+
+* This KEP is not concerned with fixing various problems with Cobra, such as third-party flags
+  leaking into the default usage or help text. We may write a future KEP to cover these issues and
+  command building in general.
+* This KEP is not concerned with the structure of flag and configuration types or their composition
+  into aggregates. We may write a future KEP to cover this.
+* This KEP is not concerned with loading and parsing config files, or resolving references contained
+  in those files (e.g. relative paths).
+* This KEP is not concerned with dynamically configuring components, or otherwise managing
+  component lifecycle.
+* This KEP is not concerned with flag or config validation.
+
+## Proposal
+
+This proposal recommends writing a wrapper package around the
+[pflag](https://godoc.org/github.com/spf13/pflag) library.
+This wrapper package is called _kflag_ in this proposal.
+
+This KEP is mostly about pinning down requirements and is accompanied by an example PR, linked
+in the below [Implementation Details](#implementation-details) section, which demonstrates _one_ way
+of accomplishing these goals.
+
+### Requirements
+
+#### General
+* Components that consume kflag should require less maintenance than components that implement
+  componentconfig without kflag.
+* kflag should be as compile-time type-safe as possible. There are many possible combinations of
+  config, so if runtime casts from interface{} are required, it may be difficult to test all
+  paths for safety.
+
+#### Flag Registration
+* Flags registered by third-party libraries must not appear in a component's command-line interface
+  unless the component owners explicitly allow it.
+* kflag should provide helpers for importing globally-registered third-party flags into a FlagSet.
+
+#### Flag Parsing
+* Components should only have to parse their command-line once, even if they have to apply the flag
+  values multiple times, or to multiple structs. The Kubelet's approach to enforcing flag-precedence
+  involves re-parsing the command-line, which led to significant complexity.
+
+#### Application of Parsed Flag Values
+* Components must be able to apply the parsed values an arbitrary number of times without
```

Comment:
I think this is covered by "Components should be able to apply the parsed flag values to structs that only represent a subset of all flag values. For instance, if a component has a Flags struct and a Config struct, it may need to apply values to each independently."

The way kflag works, you can parse everything once up-front, and pull that data out later as necessary.